lib/Target/RISCV/RISCVLDBackend.cpp

```diff
@@ -319,23 +319,35 @@ bool RISCVLDBackend::doRelaxationLui(Relocation *reloc, Relocator::DWord G) {
   size_t SymbolSize = reloc->symInfo()->outSymbol()->size();
   Relocator::DWord S = getSymbolValuePLT(*reloc);
   Relocator::DWord A = reloc->addend();
+  Relocator::DWord Value = S + A;
   uint64_t offset = reloc->targetRef()->offset();
   Relocation::Type type = reloc->type();
 
+  // First, try zero-page relaxation. It's the cheapest and does not need GP.
+
+  // Do not relax complete zeroes because they can be mistaken for
+  // not-yet-assigned values.
+  bool canRelaxZero = config().options().getRISCVRelax() &&
+                      llvm::isInt<12>(Value) && Value != 0;
+
   // HI will be deleted, LO will be converted to use GP as base.
   // GP must be available and relocation must fit in 12 bits relative to GP.
   // There is no GP for shared objects.
   bool canRelaxToGP =
       config().options().getRISCVRelax() &&
       config().options().getRISCVGPRelax() && !config().isCodeIndep() &&
-      G != 0 && fitsInGP(G, S + A, frag, reloc->targetSection(), SymbolSize);
+      G != 0 && fitsInGP(G, Value, frag, reloc->targetSection(), SymbolSize);
```

Comment:
Aside: I think this needs a `Value != 0`, in the case where `GP` is within range of zero, and something doesn't yet have an address.

Maybe we should just exit early from the whole function if `Value == 0`, on the basis that we cannot be sure it has its final address?